Source/Tests/Data/Expected/Transform_VS.msl

```diff
@@ -21,7 +21,7 @@ struct main0_in
 vertex main0_out main0(main0_in in [[stage_in]], constant type_cbVS& cbVS [[buffer(0)]])
 {
     main0_out out = {};
-    out.gl_Position = in.in_var_POSITION * cbVS.wvp;
+    out.gl_Position = cbVS.wvp * in.in_var_POSITION;
```

Comment:
I still have a little concern about the matrix. My intention is on CPU side, the caller need to transpose a matrix before putting it to cbuffer, and writing the shader in mul(vec, matrix) way. It gets consistent with D3D and HLSL. My engine hasn't get that far to enable to test the translated GLSL. Could you explain me a little more about how to use the matrix in CPU side? Thanks.

Reply:
Hi @gongminmin,

I recommend you check out Fabian Giesen's [post ](http://www.kiransprojects.com/blog/2014/row-vs-column/)about row-major and column-major matrix layouts and this Lei Zhang [post ](https://antiagainst.github.io/post/hlsl-for-vulkan-matrices/) about how to translate HLSL matrices to Vulkan.

Column-major and Row major indicate how the matrix is stored in memory, for example
a11 a12 a13
a21 a22 a23
a31 a32 a33

**Row major** is stored in memory as a11, a12, a13, a21, a22, a23, a31, a32, a33
**Column-major** is stored in memory as a11, a21, a31, a12, a22, a32, a13, a23, a33

If you want to convert from a layout to other, only you need to transpose the matrix.

Then [HLSL ](https://docs.microsoft.com/en-us/windows/win32/direct3dhlsl/dx-graphics-hlsl-per-component-math#matrix-ordering) use column-major by default, but GLSL use row-major by default.

DirectX compiler allows you specify:
**-Zpc**: Packs matrices in column-major order by default
**-Zpr**: Packs matrices in row-major order by default

I working on [WaveEngine](https://waveengine.net/) and we use HLSL as main shader languages so we store matrices as row-major in CPU and send its directly to DirectX without transpose but use -Zpr to specify row-major by default in the DirectX compiler (dxc), then we have:

**HLSL**
```
cbuffer Matrices : register(b0)
{
	float4x4 worldViewProj;
};
```

**GLSL**
```
layout(binding = 0, std140) uniform type_Matrices
{
    mat4 worldViewProj;
} Matrices;
```

The layout doesn't need to be specified, either you're using HLSL or GLSL. 

Note. You can try WaveEngine Effect Editor to convert automatically your HLSL shader to other languages (taking advantage of ShaderConductor library)

Reply:
Got it. Thanks very much for your explanation.